### PR TITLE
Refactor IdBase to provide CRTP-based printing

### PR DESCRIPTION
**Pull request by jonmeow**
This removes a lot of boilerplate `Print` functions in favor of a CRTP-based approach that uses a `Label` field as an automatic prefix. This `Label` is also made available for other purposes, particularly `IdKind` crash messages in this change. In particular, for `RequireIdKind` in  node_stack.h from using numeric IdKinds (e.g., 5 and 24) to something that will print `IdKind(<label>)` (this came up recently on #toolchain).

While I'm in here, also doing some other tinkering:

- Moving operators to be `friend` members, to reduce the extra templating now that the base types are templated.
- Adjusts IntId diagnostics from `int [...]` to `int(...)` for consistency with other id printing.
- Changes InstBlockId's label from "block" to "inst_block", since we have multiple blocks now.
- Fixes StructTypeFieldsId to use "struct_type_fields" instead of "type_block" (from `TypeBlockId`)
- Does some more adjustments from camelCase to snake_case for consistency

### CODE REVIEW

File: toolchain/check/node_stack.h [hunk 1]
```diff
@@ -690,8 +690,8 @@ class NodeStack {
       -> void {
     CARBON_CHECK(NodeKindToIdKind(parse_kind) == id_kind,
                  "Unexpected Id::Kind mapping for {0}: expected {1}, found {2}",
-                 parse_kind, static_cast<int>(id_kind),
-                 static_cast<int>(NodeKindToIdKind(parse_kind)));
+                 parse_kind, SemIR::IdKind(id_kind),
+                 SemIR::IdKind(NodeKindToIdKind(parse_kind)));
```
Comment (josh11b) [inline toolchain/check/node_stack.h:694]: It seems surprising to me that `NodeKindToIdKind` is returning an `Id::Kind` that is different from but convertible to a `SemIR::IdKind`. What is the difference? Do we need both?

File: toolchain/sem_ir/ids.cpp [hunk 2]
```diff
@@ -90,19 +94,18 @@ auto NameId::ForIdentifier(IdentifierId id) -> NameId {
 }
 
 auto NameId::Print(llvm::raw_ostream& out) const -> void {
-  out << "name";
   if (*this == SelfValue) {
-    out << "SelfValue";
+    out << Label << "SelfValue";
```
Comment (josh11b) [inline toolchain/sem_ir/ids.cpp]: This is going to be camel case as well (`nameSelfValue`, `nameSelfType`, ...). Should it be changed too?

File: toolchain/sem_ir/ids.h [hunk 3]
```diff
@@ -465,7 +438,10 @@ constexpr BoolValue BoolValue::True = BoolValue(1);
 //
 // This might eventually capture any other properties of an integer type that
 // affect its semantics, such as overflow behavior.
-struct IntKind : public IdBase, public Printable<IntKind> {
+struct IntKind : public IdBase<IntKind> {
+  // Not used by `Print`, but for `IdKind`.
+  static constexpr llvm::StringLiteral Label = "int_signed_or_unsigned";
```
Comment (josh11b) [inline toolchain/sem_ir/ids.h]: Maybe?
```suggestion
  static constexpr llvm::StringLiteral Label = "int_kind";
```

File: toolchain/sem_ir/ids.h [hunk 4]
```diff
@@ -480,16 +456,21 @@ struct IntKind : public IdBase, public Printable<IntKind> {
 constexpr IntKind IntKind::Unsigned = IntKind(0);
 constexpr IntKind IntKind::Signed = IntKind(1);
 
-// A float kind value
-struct FloatKind : public IdBase, public Printable<FloatKind> {
+// A float kind value.
+struct FloatKind : public IdBase<FloatKind> {
+  // Not used by `Print`, but for `IdKind`.
+  static constexpr llvm::StringLiteral Label = "float";
```
Comment (josh11b) [inline toolchain/sem_ir/ids.h]: To be consistent with my suggestion for `IntKind`:
```suggestion
  static constexpr llvm::StringLiteral Label = "float_kind";
```

File: toolchain/check/node_stack.h (hunk 1 above)
Comment (jonmeow) [reply toolchain/check/node_stack.h:694]: I'm pretty sure the use of RawEnumType is done with a focus on templates. We can't pass class instances as template parameters.

File: toolchain/sem_ir/ids.cpp (hunk 2 above)
Comment (jonmeow) [reply toolchain/sem_ir/ids.cpp]: Added parentheses. Also on `type`

File: toolchain/sem_ir/ids.h (hunk 3 above)
Comment (jonmeow) [reply toolchain/sem_ir/ids.h]: Done

File: toolchain/sem_ir/ids.h (hunk 4 above)
Comment (jonmeow) [reply toolchain/sem_ir/ids.h]: Done

**Review by josh11b — APPROVED** [review]

**Pull request merged**